File: src/features/desktop-monitor/SessionMonitor.ts
import { Shell } from '../../utility/Shell';
import { Logger } from '../../utility/Logger';
import type { SessionInfo } from './types';

export class SessionMonitor {
	private static instance: SessionMonitor;
	private logger = Logger.getInstance();

	private constructor() {}

	static getInstance(): SessionMonitor {
		if (!SessionMonitor.instance) {
			SessionMonitor.instance = new SessionMonitor();
		}
		return SessionMonitor.instance;
	}

	async getSessionInfo(): Promise<SessionInfo> {
		try {
			const sessionId = process.env.XDG_SESSION_ID || '';
			if (!sessionId) {
				this.logger.warn('XDG_SESSION_ID not set, session monitoring unavailable');
				return this.getDefaultSessionInfo();
			}

			const result = await Shell.execute(`loginctl show-session ${sessionId}`, {
				timeout: 5000,
			});

			if (result.exitCode !== 0) {
				this.logger.error(`loginctl failed: ${result.stderr}`);
				return this.getDefaultSessionInfo();
			}

			return this.parseLoginctlOutput(result.stdout);
		} catch (error) {
			this.logger.error(`Error getting session info: ${error}`);
			return this.getDefaultSessionInfo();
		}
	}

	private parseLoginctlOutput(output: string): SessionInfo {
		const lines = output.split('\n');
		const props: Record<string, string> = {};

		for (const line of lines) {
			const match = line.match(/^([^=]+)=(.*)$/);
			if (match) {
				props[match[1].trim()] = match[2].trim();
			}
		}

		return {
			sessionId: props.Id || process.env.XDG_SESSION_ID || 'unknown',
			user: props.Name || process.env.USER || 'unknown',
			seat: props.Seat || 'seat0',
			type: (props.Type || process.env.XDG_SESSION_TYPE || 'unknown').toLowerCase(),
			state: props.State || 'unknown',
			active: props.Active === 'yes',
			idle: props.IdleHint === 'yes',
			locked: props.LockedHint === 'yes',
			vt: parseInt(props.VTNr || '0', 10),
			display: props.Display || process.env.DISPLAY || '',
		};
	}

	private getDefaultSessionInfo(): SessionInfo {
		return {
			sessionId: process.env.XDG_SESSION_ID || 'unknown',
			user: process.env.USER || 'unknown',
			seat: 'seat0',
			type: (process.env.XDG_SESSION_TYPE || 'unknown').toLowerCase(),
			state: 'unknown',
			active: false,
			idle: false,
			locked: false,
			vt: 0,
			display: process.env.DISPLAY || '',
		};
	}

	async isSessionLocked(): Promise<boolean> {
		try {
			const info = await this.getSessionInfo();
			return info.locked;
		} catch (error) {
			this.logger.error(`Error checking lock status: ${error}`);
			return false;
		}
	}

	async getIdleStatus(): Promise<boolean> {
		try {
			const info = await this.getSessionInfo();
			return info.idle;
		} catch (error) {
			this.logger.error(`Error checking idle status: ${error}`);
			return false;
		}
	}

	async lockSession(): Promise<void> {
		try {
			const sessionId = process.env.XDG_SESSION_ID;
			if (!sessionId) {
				throw new Error('XDG_SESSION_ID not set');
			}

			const result = await Shell.execute(`loginctl lock-session ${sessionId}`, {
				timeout: 5000,
			});

			if (result.exitCode !== 0) {
				throw new Error(`loginctl lock-session failed: ${result.stderr}`);
			}

			this.logger.info('Session locked successfully');
		} catch (error) {
			this.logger.error(`Error locking session: ${error}`);
			throw error;
		}
	}

	async unlockSession(): Promise<void> {
		try {
			const sessionId = process.env.XDG_SESSION_ID;
			if (!sessionId) {
				throw new Error('XDG_SESSION_ID not set');
			}

			const result = await Shell.execute(`loginctl unlock-session ${sessionId}`, {
				timeout: 5000,
			});

			if (result.exitCode !== 0) {
				throw new Error(`loginctl unlock-session failed: ${result.stderr}`);
			}

			this.logger.info('Session unlocked successfully');
		} catch (error) {
			this.logger.error(`Error unlocking session: ${error}`);
			throw error;
		}
	}

	formatSessionInfo(info: SessionInfo): string {
		const lines = [
			'Session Information:',
			`  Session ID: ${info.sessionId}`,
			`  User: ${info.user}`,
			`  Seat: ${info.seat}`,
			`  Type: ${info.type}`,
			`  State: ${info.state}`,
			`  Active: ${info.active ? 'yes' : 'no'}`,
			`  Idle: ${info.idle ? 'yes' : 'no'}`,
			`  Locked: ${info.locked ? 'yes' : 'no'}`,
		];

		if (info.vt > 0) {
			lines.push(`  VT: ${info.vt}`);
		}

		if (info.display) {
			lines.push(`  Display: ${info.display}`);
		}

		return lines.join('\n');
	}
}

File: src/features/desktop-monitor/SessionMonitor.go
/**
 * Session monitor - monitors systemd-logind session state
 */

package desktopmonitor

import (
	"context"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/ln64-git/daemira/src/utility"
)

// SessionMonitor monitors systemd-logind session state
type SessionMonitor struct {
	logger *utility.Logger
	shell  *utility.Shell
	mu     sync.RWMutex
}

var (
	sessionMonitorInstance *SessionMonitor
	sessionMonitorOnce     sync.Once
)

// GetSessionMonitor returns the singleton SessionMonitor instance
func GetSessionMonitor() *SessionMonitor {
	sessionMonitorOnce.Do(func() {
		sessionMonitorInstance = &SessionMonitor{
			logger: utility.GetLogger(),
			shell:  utility.NewShell(utility.GetLogger()),
		}
	})
	return sessionMonitorInstance
}

// GetSessionInfo gets current session information
func (sm *SessionMonitor) GetSessionInfo(ctx context.Context) (*SessionInfo, error) {
	sessionID := os.Getenv("XDG_SESSION_ID")
	if sessionID == "" {
		sm.logger.Warn("XDG_SESSION_ID not set, session monitoring unavailable")
		return sm.getDefaultSessionInfo(), nil
	}

	result, err := sm.shell.Execute(ctx, fmt.Sprintf("loginctl show-session %s", sessionID), &utility.ExecOptions{
		Timeout: 5 * time.Second,
	})

	if err != nil || result.ExitCode != 0 {
		sm.logger.Error("loginctl failed: %v", err)
		return sm.getDefaultSessionInfo(), nil
	}

	return sm.parseLoginctlOutput(result.Stdout), nil
}

// parseLoginctlOutput parses loginctl output into SessionInfo
func (sm *SessionMonitor) parseLoginctlOutput(output string) *SessionInfo {
	lines := strings.Split(output, "\n")
	props := make(map[string]string)

	re := regexp.MustCompile(`^([^=]+)=(.*)$`)
	for _, line := range lines {
		matches := re.FindStringSubmatch(line)
		if len(matches) == 3 {
			props[strings.TrimSpace(matches[1])] = strings.TrimSpace(matches[2])
		}
	}

	sessionID := props["Id"]
	if sessionID == "" {
		sessionID = os.Getenv("XDG_SESSION_ID")
	}
	if sessionID == "" {
		sessionID = "unknown"
	}

	user := props["Name"]
	if user == "" {
		user = os.Getenv("USER")
	}
	if user == "" {
		user = "unknown"
	}

	seat := props["Seat"]
	if seat == "" {
		seat = "seat0"
	}

	sessionType := strings.ToLower(props["Type"])
	if sessionType == "" {
		sessionType = strings.ToLower(os.Getenv("XDG_SESSION_TYPE"))
	}
	if sessionType == "" {
		sessionType = "unknown"
	}

	state := props["State"]
	if state == "" {
		state = "unknown"
	}

	active := props["Active"] == "yes"
	idle := props["IdleHint"] == "yes"
	locked := props["LockedHint"] == "yes"

	vt := 0
	if vtStr := props["VTNr"]; vtStr != "" {
		if v, err := strconv.Atoi(vtStr); err == nil {
			vt = v
		}
	}

	display := props["Display"]
	if display == "" {
		display = os.Getenv("DISPLAY")
	}

	return &SessionInfo{
		SessionID: sessionID,
		User:      user,
		Seat:      seat,
		Type:      sessionType,
		State:     state,
		Active:    active,
		Idle:      idle,
		Locked:    locked,
		VT:        vt,
		Display:   display,
	}
}

// getDefaultSessionInfo returns default session info when loginctl fails
func (sm *SessionMonitor) getDefaultSessionInfo() *SessionInfo {
	sessionID := os.Getenv("XDG_SESSION_ID")
	if sessionID == "" {
		sessionID = "unknown"
	}

	user := os.Getenv("USER")
	if user == "" {
		user = "unknown"
	}

	sessionType := strings.ToLower(os.Getenv("XDG_SESSION_TYPE"))
	if sessionType == "" {
		sessionType = "unknown"
	}

	display := os.Getenv("DISPLAY")

	return &SessionInfo{
		SessionID: sessionID,
		User:      user,
		Seat:      "seat0",
		Type:      sessionType,
		State:     "unknown",
		Active:    false,
		Idle:      false,
		Locked:    false,
		VT:        0,
		Display:   display,
	}
}

// IsSessionLocked checks if session is locked
func (sm *SessionMonitor) IsSessionLocked(ctx context.Context) (bool, error) {
	info, err := sm.GetSessionInfo(ctx)
	if err != nil {
		return false, err
	}
	return info.Locked, nil
}

// GetIdleStatus checks if session is idle
func (sm *SessionMonitor) GetIdleStatus(ctx context.Context) (bool, error) {
	info, err := sm.GetSessionInfo(ctx)
	if err != nil {
		return false, err
	}
	return info.Idle, nil
}

// LockSession locks the current session
func (sm *SessionMonitor) LockSession(ctx context.Context) error {
	sessionID := os.Getenv("XDG_SESSION_ID")
	if sessionID == "" {
		return fmt.Errorf("XDG_SESSION_ID not set")
	}

	result, err := sm.shell.Execute(ctx, fmt.Sprintf("loginctl lock-session %s", sessionID), &utility.ExecOptions{
		Timeout: 5 * time.Second,
	})

	if err != nil || result.ExitCode != 0 {
		return fmt.Errorf("loginctl lock-session failed: %v", err)
	}

	sm.logger.Info("Session locked successfully")
	return nil
}

// UnlockSession unlocks the current session
func (sm *SessionMonitor) UnlockSession(ctx context.Context) error {
	sessionID := os.Getenv("XDG_SESSION_ID")
	if sessionID == "" {
		return fmt.Errorf("XDG_SESSION_ID not set")
	}

	result, err := sm.shell.Execute(ctx, fmt.Sprintf("loginctl unlock-session %s", sessionID), &utility.ExecOptions{
		Timeout: 5 * time.Second,
	})

	if err != nil || result.ExitCode != 0 {
		return fmt.Errorf("loginctl unlock-session failed: %v", err)
	}

	sm.logger.Info("Session unlocked successfully")
	return nil
}

// FormatSessionInfo formats session info for display
func (sm *SessionMonitor) FormatSessionInfo(info *SessionInfo) string {
	lines := []string{
		"Session Information:",
		fmt.Sprintf("  Session ID: %s", info.SessionID),
		fmt.Sprintf("  User: %s", info.User),
		fmt.Sprintf("  Seat: %s", info.Seat),
		fmt.Sprintf("  Type: %s", info.Type),
		fmt.Sprintf("  State: %s", info.State),
		fmt.Sprintf("  Active: %s", boolToYesNo(info.Active)),
		fmt.Sprintf("  Idle: %s", boolToYesNo(info.Idle)),
		fmt.Sprintf("  Locked: %s", boolToYesNo(info.Locked)),
	}

	if info.VT > 0 {
		lines = append(lines, fmt.Sprintf("  VT: %d", info.VT))
	}

	if info.Display != "" {
		lines = append(lines, fmt.Sprintf("  Display: %s", info.Display))
	}

	return strings.Join(lines, "\n")
}

// boolToYesNo converts bool to "yes"/"no"
func boolToYesNo(b bool) string {
	if b {
		return "yes"
	}
	return "no"
}
